sharding/shard.go

```diff
@@ -0,0 +1,219 @@
+package sharding
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type shardBackend interface {
+	Get(k common.Hash) ([]byte, error)
+	Has(k common.Hash) bool
+	Put(k common.Hash, val []byte)
+	Delete(k common.Hash)
+}
+
+// Shard base struct.
+type Shard struct {
+	shardDB shardBackend
+	shardID *big.Int
+}
+
+// MakeShard creates an instance of a Shard struct given a shardID.
+func MakeShard(shardID *big.Int, shardDB shardBackend) *Shard {
+	return &Shard{
+		shardID: shardID,
+		shardDB: shardDB,
+	}
+}
+
+// ShardID gets the shard's identifier.
+func (s *Shard) ShardID() *big.Int {
+	return s.shardID
+}
+
+// ValidateShardID checks if header belongs to shard.
+func (s *Shard) ValidateShardID(h *CollationHeader) error {
+	if s.ShardID().Cmp(h.ShardID()) != 0 {
+		return fmt.Errorf("collation does not belong to shard %d but instead has shardID %d", h.ShardID().Int64(), s.ShardID().Int64())
+	}
+	return nil
+}
+
+// HeaderByHash of collation.
+func (s *Shard) HeaderByHash(hash *common.Hash) (*CollationHeader, error) {
+	encoded, err := s.shardDB.Get(*hash)
+	if err != nil {
+		return nil, fmt.Errorf("header not found: %v", err)
+	}
+
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(encoded), uint64(len(encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	return &header, nil
+}
+
+// CollationByHash fetches full collation.
+func (s *Shard) CollationByHash(headerHash *common.Hash) (*Collation, error) {
+	header, err := s.HeaderByHash(headerHash)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := s.BodyByChunkRoot(header.ChunkRoot())
+	if err != nil {
+		return nil, err
+	}
+	return &Collation{header: header, body: body}, nil
+}
+
+// CanonicalCollationHash gets a collation header hash that has been set as canonical for
+// shardID/period pair
+func (s *Shard) CanonicalCollationHash(shardID *big.Int, period *big.Int) (*common.Hash, error) {
+	key := canonicalCollationLookupKey(shardID, period)
+	hash := common.BytesToHash(key.Bytes())
+	collationHashBytes, err := s.shardDB.Get(hash)
+	if err != nil {
+		return nil, fmt.Errorf("no canonical collation set for period, shardID pair: %v", err)
+	}
+	collationHash := common.BytesToHash(collationHashBytes)
```

Comment:
What happens when the data isn't there but there was no error? i.e. if `collationHashBytes == nil`